mtags/src/main/scala-2/scala/meta/internal/pc/InferredMethodProvider.scala

```diff
@@ -0,0 +1,520 @@
+                              _,
+                              _,
+                              template
+                            ) =>
+                          val insertPos: Position =
+                            inferEditPosition(template)
+                          signature(
+                            name = errorMethod.name.toString(),
+                            paramsString =
+                              argumentsString(arguments).getOrElse(""),
+                            retType = None,
+                            postProcess = method => {
+                              if (
+                                hasBody(
+                                  template.pos.source.content
+                                    .map(_.toString)
+                                    .mkString,
+                                  template
+                                ).isDefined
+                              )
+                                s"\n  $method"
+                              else s" {\n  $method}"
+                            },
+                            position = Option(insertPos)
+                          )
+                        case _ =>
+                          Nil
+                      }
+                  }
+                case _ =>
+                  Nil
+              }
+            } else
+              Nil
+          case _ => Nil
+        }
+      case _ =>
+        Nil
+    }
+    new WorkspaceEdit(
+      uriTextEditPairs
+        .groupBy(_._1)
+        .map { case (k, kvs) => (k, kvs.map(_._2).asJava) }
+        .toMap
+        .asJava
+    )
+  }
+
+  private def insertPosition(): Position = {
+    val blockOrTemplateIndex =
+      lastVisitedParentTrees.tail.indexWhere {
+        case _: Block | _: Template => true
+        case _ => false
+      }
+    lastVisitedParentTrees(blockOrTemplateIndex).pos
+  }
+
+  private def indentation(text: String, pos: Int): String =
+    if (pos > 0) {
+      val isSpace = text(pos) == ' '
+      val isTab = text(pos) == '\t'
+      val indent =
+        countIndent(params.text(), pos, 0)
+
+      if (isSpace) " " * indent else if (isTab) "\t" * indent else ""
+    } else ""
+
+  @tailrec
+  private def countIndent(text: String, index: Int, acc: Int): Int = {
+    if (text(index) != '\n') countIndent(text, index - 1, acc + 1)
+    else acc
+  }
+
+  // TODO taken from OverrideCompletion: move into a utility?
+  /**
+   * Get the position to insert implements for the given Template.
+   * `class Foo extends Bar {}` => retuning position would be right after the opening brace.
+   * `class Foo extends Bar` => retuning position would be right after `Bar`.
+   *
+   * @param text the text of the original source code.
+   * @param t the enclosing template for the class/object/trait we are implementing.
+   */
+  private def inferEditPosition(t: Template): Position = {
+    // get text via reflection because Template could be in a different file
```

Comment:
I would avoid handling separate files as it is quite complex to do. Anything dealing with multiple files gets complex pretty quickly.  The other file might be a library or a completely different module. You might be using a symbol that is not available in that different scope also. Or what happens if the file is Java? There is a lot of edge cases unfortunately. Keeping the functionality in a single file gives us much less complexity.

What about instead of going to another file we would create an implicit class providing the method if the symbol is not in the current file (that could be a separate improvement). And we could have a separate action that would export the implicit class to the actual type separately using semanticdb (anything dealing with multiple files usually uses semanticdb)

Reply:
oh I didn't think about that, well spotted!
Without that it reduces the advantage of using the action, I think.
I am fine with removing this to retain Metals stability and I will look into baking an Emacs specific hack to support this same action to act on multiple files.
The implicit idea seems alright a workaround, although I wouldn't use it myself so we can leave it as a further work.

Reply:
I feel like this is still very useful, especially in current file. For anything outside the current context it might not be that predictable. You might have mistyped a method name and the action would create that mistyped in a totally different class, which you would not be that aware of.

But if you are very keen on this we could do it another way:
- find the definition path for the class/object symbol
- take type information about parameter types from semanticdb 
- invoke the inferredMethod on the file the type is from, give it that type data plus the name of the method and use presentation compiler in that file to create it.

That is more complex though unfortuntely.

Reply:
For this PR we could fist find definition and if it's the same file invoke the action otherwise not do that. This gives us the window to implement the second part at some time later on.

Reply:
It would also be worth seeing how other editors or language servers do it, but for now having a smaller, surer scope seems a way to go.

Reply:
> For this PR we could fist find definition and if it's the same file invoke the action otherwise not do that. This gives us the window to implement the second part at some time later on.

indeed, I can just check the params.uri and the Template.symbol.source.uri match and print something like unsupported otherwise.

Reply:
addressed in https://github.com/scalameta/metals/pull/6877/commits/a96e7ab4f844b4d0c9695bb263876ad231ad190a